src/tactic/polyrith.lean

```diff
@@ -275,34 +276,44 @@ ch '('
     <|> sub_parser poly_parser <|> mul_parser poly_parser <|> pow_parser poly_parser)
   <* ch ')'
 
+meta instance : non_null_json_serializable poly :=
+{ to_json := λ p, json.null,  -- we don't actually need this, but the typeclass asks for it
+  of_json := λ j, do
+    s ← of_json string j,
+    match poly_parser.run_string s with
+    | sum.inl s := exceptional.fail format!"unable to parse polynomial from.\n\n{s}"
+    | sum.inr p := pure p
+    end}
+
+/-- A schema for success messages from the python script -/
+@[derive [non_null_json_serializable, inhabited]]
+structure sage_json_success :=
+(success : {b : bool // b = tt})
+(trace : option string := none)
+(data : option (list poly) := none)
+
+/-- A schema for failure messages from the python script -/
+@[derive [non_null_json_serializable, inhabited]]
+structure sage_json_failure :=
+(success : {b : bool // b = ff})
```

Comment:
I trust this is right (also because it works), but what's the significance of this weird subtype? Is the idea that only json data with `success: false` should be parsed into this type? Is the `non_null_json_serializable` derive handler the part that's taking this into account?

Reply:
The `subtype.non_null_json_serializable` parser is the one that handles this; it parses the value, and uses decidability for the proof.

I could also have used
```suggestion
(success : bool)
(success_eq_ff : success = ff)
```
And then the derive handler would have handled it; but then the `inhabited` derive handler wouldn't have worked, not that that really matters.